btree/btree.c

```diff
@@ -2415,21 +2433,25 @@ static struct m0_btree_oimpl *level_alloc(int height)
 	return oi;
 }
 
-static void level_cleanup(struct m0_btree_oimpl *oi)
+static void level_cleanup(struct m0_btree_oimpl *oi,
+			  enum m0_btree_opcode bo_opc)
 {
 	int i;
-	for (i = 0; i < oi->i_used; ++i) {
+	for (i = 0; i <= oi->i_used; ++i) {
 		if (oi->i_level[i].l_node != NULL) {
 			node_put(oi->i_level[i].l_node);
 			oi->i_level[i].l_node = NULL;
 		}
-		if (oi->i_level[i].l_alloc != NULL) {
-			node_put(oi->i_level[i].l_alloc);
-			oi->i_level[i].l_alloc = NULL;
-		}
 	}
-	if (oi->i_extra_node != NULL) {
-		node_put(oi->i_extra_node);
+	if (bo_opc == M0_BO_PUT) {
+		for (i = 0; i <= oi->i_used; ++i) {
+			if (oi->i_level[i].l_alloc != NULL) {
+				if(oi->i_level[i].l_alloc_flag)
```

Comment:
This flag l_alloc_flag is set values NODE_USED or NODE_FREE which are both non-zero so this test will always be TRUE. So this check does not make sense.
Also there is no need to add a flag such as l_alloc_flag. Instead the l_alloc pointer is normall set to NULL by default. When any possible sibling nodes are allocated then this pointer will have a non-zero value. Once the assigned value of the node is connected to the tree then the l_alloc pointer can be reset to NULL (after releasing the descriptor) so that if this variable l_alloc has a non-zero value in level_cleanup we are assured that the allocated node was not used and hence need to be freed in the level_cleanup routine.

Reply:
Okay I will make changes